include/bout/petsc_interface.hxx

```diff
@@ -0,0 +1,576 @@
+    yoffset = 0;
+  }
+
+  /// Copy constructor
+  PetscMatrix(const PetscMatrix<F>& m) : matrix(new Mat(), MatrixDeleter()), pt(m.pt) {
+    MatDuplicate(*m.matrix, MAT_COPY_VALUES, matrix.get());
+    // MatCopy(*m.matrix, *matrix);
+    indexConverter = m.indexConverter;
+    yoffset = m.yoffset;
+    initialised = m.initialised;
+  }
+  
+  /// Move constrcutor
+  PetscMatrix(PetscMatrix<F>&& m) : pt(m.pt) {
+    matrix = m.matrix;
+    indexConverter = m.indexConverter;
+    yoffset = m.yoffset;
+    initialised = m.initialised;
+    m.initialised = false;
+  }
+
+  // Construct a matrix capable of operating on the specified field
+  PetscMatrix(F &f) : matrix(new Mat(), MatrixDeleter()) {
+    MPI_Comm comm;
+    if (std::is_same<F, FieldPerp>::value) {
+      comm = f.getMesh()->getXcomm();
+    } else {
+      comm = BoutComm::get();
+    }
+    indexConverter = GlobalIndexer::getInstance(f.getMesh());
+    pt = &f.getMesh()->getCoordinates()->getParallelTransform();
+    int size;
+    if (std::is_same<F, FieldPerp>::value) {
+      size = f.getMesh()->localSizePerp();
+    } else if (std::is_same<F, Field2D>::value) {
+      size = f.getMesh()->localSize2D();
+    } else if (std::is_same<F, Field3D>::value) {
+      size = f.getMesh()->localSize3D();
+    } else {
+      throw BoutException("PetscVector initialised for non-field type.");
+    }
+    MatCreate(comm, matrix.get());
+    MatSetSizes(*matrix, size, size, PETSC_DECIDE, PETSC_DECIDE);
+    MatSetType(*matrix, MATMPIAIJ);
+    MatSetUp(*matrix);
+    yoffset = 0;
+    initialised = true;
+  }
+  
+  /// Copy assignment
+  PetscMatrix<F>& operator=(const PetscMatrix<F>& rhs) {
+    swap(*this, rhs);
+    return *this;
+  }
+  /// Move assignment
+  PetscMatrix<F>& operator=(PetscMatrix<F>&& rhs) {
+    matrix = rhs.matrix;
+    indexConverter = rhs.indexConverter;
+    pt = rhs.pt;
+    yoffset = rhs.yoffset;
+    initialised = rhs.initialised;
+    rhs.initialised = false;
+  }
+  friend void swap<F>(PetscMatrix<F>& first, PetscMatrix<F>& second);
+
+  PetscMatrixElement& operator()(ind_type& index1, ind_type& index2) {
+    int global1 = indexConverter->getGlobal(index1),
+        global2 = indexConverter->getGlobal(index2);
+#if CHECKLEVEL >= 1
+    if (!initialised) {
+      throw BoutException("Can not return element of uninitialised matrix");
+    } else if (global1 == -1 || global2 == -1) {
+      throw BoutException("Request to return invalid matrix element");
+    }
+#endif
+    std::vector<PetscInt> positions;
+    std::vector<PetscScalar> weights;
+    if (yoffset != 0) {
+      std::vector<ParallelTransform::positionsAndWeights> pw;
+      if (yoffset == -1) {
+	pw = pt->getWeightsForYDownApproximation(index2.x(), index2.y(), index2.z());
+      } else if (yoffset == 1) {
+	pw = pt->getWeightsForYUpApproximation(index2.x(), index2.y(), index2.z());
+      } else {
+	pw = pt->getWeightsForYApproximation(index2.x(), index2.y(), index2.z(), yoffset);
+      }
```

Comment:
An interesting technique to make `pw` `const` here is the "immediately invoked function expression", or IIFE:

```cpp
const auto pw = [yoffset, &index2]() {
      if (yoffset == -1) {
	return = pt->getWeightsForYDownApproximation(index2.x(), index2.y(), index2.z());
      } else if (yoffset == 1) {
	return = pt->getWeightsForYUpApproximation(index2.x(), index2.y(), index2.z());
      } else {
	return = pt->getWeightsForYApproximation(index2.x(), index2.y(), index2.z(), yoffset);
      }
}();
```

Reply:
I'm happy to make the change, but what exactly is the advantage of this?

Reply:
Being able to make the local variable `const`. The advantage to making things `const` is that it signals to the reader that they don't need to look elsewhere for modifications to the value: it's set here and not touched again later.

It's plausible that the compiler can make some optimisations too, although usually not.

Usually you don't need quite such ugly constructions as IIFEs, but generally prefer to make local variables `const` if you can.